cpp/src/arrow/adapters/orc/adapter.cc

```diff
@@ -0,0 +1,698 @@
+      auto fields = std::vector<std::shared_ptr<Field>>();
+      std::vector<uint8_t> type_codes;
+      for (uint64_t child = 0; child < size; ++child) {
+        std::shared_ptr<DataType> elemtype;
+        RETURN_NOT_OK(get_dtype(type->getSubtype(child), &elemtype));
+        auto f = field("_union_" + std::to_string(child), elemtype);
+        fields.push_back(f);
+        type_codes.push_back((uint8_t)child);
+      }
+      *out = union_(fields, type_codes);
+      break;
+    }
+    default: {
+      std::stringstream ss;
+      ss << "Unknown Orc type kind: " << kind;
+      return Status::Invalid(ss.str());
+    }
+  }
+  return Status::OK();
+}
+
+class ORCFileReader::Impl {
+ public:
+  explicit Impl(MemoryPool* pool, std::unique_ptr<liborc::Reader> reader)
+      : pool_(pool), reader_(std::move(reader)) {}
+
+  static Status Open(const std::shared_ptr<io::ReadableFileInterface>& file,
+                     MemoryPool* pool, std::unique_ptr<Impl>* impl) {
+    std::unique_ptr<ArrowInputFile> io_wrapper(new ArrowInputFile(file));
+    liborc::ReaderOptions options;
+    std::unique_ptr<liborc::Reader> liborc_reader;
+    try {
+      liborc_reader = createReader(std::move(io_wrapper), options);
+    } catch (const liborc::ParseError& e) {
+      return Status::IOError(e.what());
+    }
+    impl->reset(new Impl(pool, std::move(liborc_reader)));
+    RETURN_NOT_OK((*impl)->Init());
+
+    return Status::OK();
+  }
+
+  Status Init() {
+    int64_t nstripes = reader_->getNumberOfStripes();
+    stripes_.resize(nstripes);
+    std::unique_ptr<liborc::StripeInformation> stripe;
+    for (int i = 0; i < nstripes; ++i) {
+      stripe = reader_->getStripe(i);
+      stripes_[i] = StripeInformation(
+          {stripe->getOffset(), stripe->getLength(), stripe->getNumberOfRows()});
+    }
+    return Status::OK();
+  }
+
+  uint64_t NumberOfStripes() { return stripes_.size(); }
+
+  uint64_t NumberOfRows() { return reader_->getNumberOfRows(); }
+
+  Status ReadSchema(std::shared_ptr<Schema>* out) {
+    const liborc::Type& type = reader_->getType();
+    return type_to_schema(type, out);
+  }
+
+  Status type_to_schema(const liborc::Type& type, std::shared_ptr<Schema>* out) {
+    if (type.getKind() != liborc::STRUCT) {
+      return Status::NotImplemented(
+          "Only ORC files with a top-level struct "
+          "can be handled");
+    }
+    auto size = type.getSubtypeCount();
+    auto fields = std::vector<std::shared_ptr<Field>>();
+    for (uint64_t child = 0; child < size; ++child) {
+      std::shared_ptr<DataType> elemtype;
+      RETURN_NOT_OK(get_dtype(type.getSubtype(child), &elemtype));
+      std::string name = type.getFieldName(child);
+      fields.push_back(field(name, elemtype));
+    }
+    std::list<std::string> keys = reader_->getMetadataKeys();
+    std::shared_ptr<KeyValueMetadata> metadata = nullptr;
+    if (!keys.empty()) {
+      metadata = std::make_shared<KeyValueMetadata>();
```

Comment:
This is not needed, the value of metadata is already this.

Reply:
Are you sure about this? Removing this line leads to a segfault, as the `metadata->Append` line below dereferences a nullptr (IIUC, note that my c++ is rusty at best, so I may misunderstand here).

Reply:
@xhochy This is needed if you don't pass the keys and values to the metadata constructor, because a NULL pointer to `KeyValueMetadata` is different than an instance of `KeyValueMetadata` with no keys/values.